### ISSUE TITLE
Q-Menu gets squashed and shows scrollbar if it does not fit to the right and to the bottom both.

### ISSUE THREAD

**Issue by BarnabasSzabolcs** [quasarframework/quasar#8554]
**Describe the bug**
I use the q-menu component as a contextual menu, showing an "advanced" content (not only a list) 
that sometimes does not fit where the menu would be normally located (top-left to target's bottom-left),
and if the right side of the target and the bottom side of the target does not have enough space at the same time,
the q-menu does not only get relocated which works nicely, but the window also gets squashed and the contents appear with a scrollbar.

**Codepen/jsFiddle/Codesandbox (required)**
Fork a Codepen (https://codepen.quasar.dev) or a jsFiddle (https://jsfiddle.quasar.dev) or a Codesandbox (https://codesandbox.quasar.dev) and hit save then copy-paste link here.

https://codepen.io/barnab-s-szabolcs/pen/dyOmeWd

**To Reproduce**
1. Go to '...' the pen
2. Click on '....'  the blue button
3. See error (if the context menu can appear under the target, everything's fine. if not, the context menu gets scrolled, even if it has enough space not to scroll) 

**Expected behavior**
The q-menu appears without scroll, in the displaced location.

**Screenshots**

**Platform (please complete the following information):**
Quasar Version:
@quasar/app Version:
Quasar mode:
  [ x ] SPA
  [ ] SSR
  [ ] PWA
  [ ] Electron
  [ ] Cordova
  [ ] Capacitor
  [ ] BEX
Tested on:
  [ x ] SPA
  [ ] SSR
  [ ] PWA
  [ ] Electron
  [ ] Cordova
  [ ] Capacitor
  [ ] BEX
OS: macOS High Sierra 10.13.6 (17G65)
Node: v14.4.0
NPM: 6.14.4
Yarn: 1.21.1
Browsers: Firefox (Interestingly in Chrome the scrollbar first appears then the q-menu resizes itself and the scrollbar disappears)
iOS:
Android:
Electron:

**Additional context**

**Comment by BarnabasSzabolcs** [comment]
![Screen Shot 2021-02-28 at 22 29 44](https://user-images.githubusercontent.com/2684538/109434246-a4ef1a80-7a14-11eb-8bb0-245e508ccfa6.png)

**Comment by pdanpdan** [comment]
Hello, can you please test if this fixes the problem for you?

In package.json replace quasar dependency with:
"quasar": "https://github.com/pdanpdan/quasar#quasar-pdan-v1.15.4-beta.1"

Then do a yarn or npm install and start the app again.

Please report if you find something not working.

**Comment by BarnabasSzabolcs** [comment]
Multumesc Dan, 
now the scrolling is gone!😎
The only issues that are left:
1. the q-menu window became "jumpy" (first appears at the wrong location then jumps into its correct location). 
2. the min-width parameter of q-menu is not respected, unless I apply min-width to an inner div.

**Comment by BarnabasSzabolcs** [comment]
Hmmm. Maybe it would be worth to make this automatic behaviour disablable. (or force-on-screen="" a/y/x/t/b/l/r)  
If a context is scrollable, the context menu scrolls nicely along with the content, so the user can just scroll down,
and the benefit of this is that the user can expect the context menu always at the same location.

**Comment by pdanpdan** [comment]
I cannot reproduce the case with `imprecise`. From the image it looks like there is no space on the right (so it switches to left) but the space on bottom should be more than enough.
I can only think that the anchor element (or some parent) is fixed positioned - in that case the menu will stay inside the screen, not inside the document.

**Comment by BarnabasSzabolcs** [comment]
You may be right, the q-menu is within the contents of a q-scroll-area that is eventually within a q-drawer.
And even though the space on the bottom should be more than enough, it likes to get displayed up, and sometimes it even jumps down after appearing up first. It must be pretty difficult to determine correct positioning for a context menu in a very general case.

**Comment by pdanpdan** [comment]
If you could prepare a reproduction case for this i can tweak the algorithms a little.
Btw, if you use beta.2 you'll have min-width/height

**Comment by BarnabasSzabolcs** [comment]
https://user-images.githubusercontent.com/2684538/109825802-75b9f280-7c3a-11eb-9e0a-99e599cb0868.mov

**Comment by pdanpdan** [comment]
https://pdanpdan.github.io/quasar-docs/start/umd#Installation

**Comment by BarnabasSzabolcs** [comment]
Cool cool, I've updated the pen, https://codepen.io/barnab-s-szabolcs/pen/dyOmeWd
Now, try and set the pen's window 
- 1114×673 - here the q-menu scrolls
- 717×703 - here the up-down jump is observable. 

The pen uses:
https://cdn.jsdelivr.net/npm/vue/dist/vue.min.js  ( vue@^2.0.0 didn't get accepted by codepen but this vue is 2.6.12 anyway)
https://cdn.jsdelivr.net/gh/pdanpdan/quasar@quasar-pdan-v1.15.4-beta.2/dist/quasar.umd.min.js
https://cdn.jsdelivr.net/npm/@quasar/extras/material-icons/material-icons.css
https://cdn.jsdelivr.net/gh/pdanpdan/quasar@quasar-pdan-v1.15.4-beta.2/dist/quasar.min.css

**Comment by pdanpdan** [comment]
You can try beta.3 - it should improve on the initial positioning.

If the content of the dialog takes some time to be rendered (and on the initial render it is smaller than the final one) then the dialog might end up in the requested position and with scroll bars - you can mitigate this with a min-width/height.

About the positioning when the anchor is inside a scrollable element (other than the document.scrollingElement) - it is too complex to compute, I'm not sure it's worth, but I'll think about it more.

**Comment by BarnabasSzabolcs** [comment]
Yes, I think you're right. q-menu is meant to be a toolbar menu anyways.

I'm not sure if you've also discovered, that on mobile if the menu is opened from the center
and it cannot align itself either left or right, it will break the layout by expanding leftwards.
I don't think that is a huge issue for most users though.

**Comment by pdanpdan** [comment]
@BarnabasSzabolcs Sorry, I was busy lately. What do you mean by
> I have checked the context menus on iPhone mobile and if the menu is opened from the center
and it cannot align itself either left or right, **it will break the layout by expanding leftwards**

**Comment by BarnabasSzabolcs** [comment]
> @BarnabasSzabolcs Sorry, I was busy lately. What do you mean by
> 
> > I have checked the context menus on iPhone mobile and if the menu is opened from the center
> > and it cannot align itself either left or right, **it will break the layout by expanding leftwards**
> 
> Btw, you can use `"quasar": "https://github.com/pdanpdan/quasar#quasar-pdan-v1.15.5-beta.17"` :)

@pdanpdan no worries at all, let me illustrate with iphone screenshots:
All the examples down here are correct except for *two-time* that also causes the phone to enable scrolling horizontally. The fixed sections of the layout stay in place, only the main section "breaks".

Funnily enough, *teatime*'s menu is displayed correctly at the left side of the screen, even though it does not fit to the left or the right.

Please not that the screenshots were made with 1.15.4-beta.15. 

I could see there were many developments with 1.15.5-beta.17! (I cannot yet use beta.17 because there I have a vertical scrollbar appearing at my vertical flex layout so I'll have to investigate that first. That's not related to this q-menu issue though.)

![1](https://user-images.githubusercontent.com/2684538/111542787-79896100-8772-11eb-8803-984b80dfcce8.jpg)
![2](https://user-images.githubusercontent.com/2684538/111542790-7aba8e00-8772-11eb-962f-60413162831c.jpg)
![3](https://user-images.githubusercontent.com/2684538/111542794-7b532480-8772-11eb-82d3-f111f9f9160b.jpg)
![4](https://user-images.githubusercontent.com/2684538/111542799-7bebbb00-8772-11eb-9cc5-176a60e0ca61.jpg)
![5](https://user-images.githubusercontent.com/2684538/111542801-7c845180-8772-11eb-97c3-7f840a8b26c4.jpg)
![6](https://user-images.githubusercontent.com/2684538/111542804-7d1ce800-8772-11eb-9b0c-822e4bf2aa6a.jpg)

**Comment by rstoenescu** [comment]
Closing after a retest against the current Quasar v2 (`dev`, post v2.28.0).

The positioning engine now flips each axis independently and only caps the popup's size when it does not fit the roomier side either. A menu on an anchor at the bottom-right corner opens above and to the left, flush with the anchor's top and right edges, at its natural size and with no scrollbars. This holds on both the CSS anchor positioning engine (Chromium) and the JS fallback (Firefox, Safari), and it is covered by tests. The placement is also settled before the popup becomes visible, so the initial jump discussed in the thread is gone.

If you still see a squashed or scrolled menu on a current Quasar v2 release, please open a new issue with a reproduction against that version.
